tests/Tests/Protostar.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeOperators       #-}
+
+module Tests.Protostar (specProtostar) where
+
+import           Control.Monad                               (replicateM)
+import           Data.Kind                                   (Type)
+import           Prelude                                     (IO, id, type (~), ($), (.), (<$>), (<*>), (<>))
+import qualified Prelude                                     as P
+import qualified Test.Hspec
+import           Test.Hspec                                  (Spec, describe, hspec)
+import           Test.QuickCheck
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Field
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Algebra.EllipticCurve.BLS12_381
+import           ZkFold.Base.Algebra.EllipticCurve.Class
+import qualified ZkFold.Base.Data.Vector                     as V
+import           ZkFold.Base.Data.Vector                     (Vector)
+import           ZkFold.Base.Protocol.Protostar
+import           ZkFold.Prelude                              ((!!))
+import           ZkFold.Symbolic.Class
+import           ZkFold.Symbolic.Compiler
+import           ZkFold.Symbolic.Data.FieldElement           (FieldElement)
+
+data RecursiveFunction n c a
+    = RecursiveFunction
+        { rIterations :: Natural
+        , rInitial    :: Vector n a
+        , rFunction   :: Vector n (FieldElement c) -> Vector n (FieldElement c)
+        }
+
+instance P.Show a => P.Show (RecursiveFunction n c a) where
+    show RecursiveFunction{..} = P.unlines [P.show rIterations, P.show rInitial]
+
+instance
+    ( KnownNat n
+    , Arbitrary a
+    , Symbolic c
+    , MultiplicativeSemigroup (FieldElement c)
```

Comment:
This should be inferrable from `Symbolic c` constraint

Reply:
Yeah, there are a lot of cases where GHC does not flag redundant constraints. In some cases where constraint relations were not obvious I even had to remove constraints one-by-one and observe if it would lead to errors